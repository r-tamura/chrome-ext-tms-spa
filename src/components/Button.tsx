import * as React from "react"
import classNames from "~/helpers/classNames"
import { Link } from "react-router-dom"

enum Color {
  DEFAULT          = "",
  PRIMARY          = "primary",
  SECONDARY        = "secondary",
  DANGER           = "danger",
  LINK             = "link",
}

interface IButtonProps extends React.ButtonHTMLAttributes<{}> {
  /* button要素とするかa要素とするか true: button / false: a*/
  button?: boolean
  className?: string
  color?: Color
  /* ボタンをblock要素とするか コンテナコンポーネントに対して最大幅と同じになる */
  block?: boolean
  circle?: boolean
  component?: React.AllHTMLAttributes<{}>
  to?: string
  size?: string
}

const Button: React.SFC<IButtonProps> = ({
  button        = true,
  disabled      = false,
  color         = Color.DEFAULT,
  block         = false,
  circle        = false,
  className: classNameProps,
  component,
  children,
  ...otherProps
} = {}) => {
  const className = classNames(
    button ? "tms-btn" : null,
    color || Color.DEFAULT,
    { disabled, block, circle },
    classNameProps
  )
  const Component = component || button ? "button" : Link

  return (
    <Component className={className} {...otherProps}>
      {children}
    </Component>
  )
}

export {
  IButtonProps,
  Color,
  Button,
}

export default Button
