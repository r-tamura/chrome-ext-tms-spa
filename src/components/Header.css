:root {
  --header-height: 50px;
}

.company-logo {
  display: block;
  width: 40px;
}


.page-header {
  display: flex;
  align-items: center;
  flex-direction: row;
  padding: 0 5px;
  background-color: var(--primary-color-reverse);
  /* color: var(--white); */
  font-size: var(--font-size-lg);
  font-weight: bolder;
  line-height: var(--header-height);
  
  @apply --z-index-high;
  @apply --box-shadow;

  & a:focus, & a:hover {
    opacity: 0.8
  }

  & .nav {
    height: 80%;
    margin: 5px 0;
  }

  & .nav-link {
    height: 100%;

    &.active {
      background-color: $nav-selected-bg-color;
    }
  }

  @media (--md) {
    /* ロゴ非表示 */
    .nav, .nav-link {
        display: none;
    }

    .menu-btn {
        display: block;
    }
  }
}

/* ヘッダーメニューボタン */
.menu-btn {
    cursor: pointer;
    padding: $header-height/4 10px;
    display: none;
    &:hover {
      background-color: darken($COLOR_HEADER_BG, 10%);
      border-radius: 50%;
    }

    img {
      height: 50%;
    }
}

.header-user {
  /* @extend .clickable; */
  display: flex;
  align-items: center;
  margin-left: auto;

  .person-icon {
    background-image: url(chrome-extension://odbcdehjjdjjnpenjbblpmkkibggblpe/chrome/assets/images/person.svg);
    height: 50%;
    width: 25px;
    display: block;
  }

  &:hover {
    text-decoration: underline;
  }
}

.header-user, .nav-link {
  img {
    display: inline-block;
    height: $header-height;
    width: $header-height;
    padding: $header-height * 0.10 10px;
  }
}

.header-user .dropdown-menu {
  right: 0px;
}
