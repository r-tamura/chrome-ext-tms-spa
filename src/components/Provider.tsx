import React from "react";
import { ThemeProvider } from "styled-components";
import { telema } from "~/styles/theme";

export const AppThemeProvider: React.SFC<{ children: JSX.Element }> = ({
  children
}) => <ThemeProvider theme={telema}>{children}</ThemeProvider>;
