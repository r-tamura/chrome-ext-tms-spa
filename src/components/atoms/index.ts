export * from "./Button";
export * from "./Panel";
export * from "./Modal";
export * from "./Form";
export * from "./TextField";
