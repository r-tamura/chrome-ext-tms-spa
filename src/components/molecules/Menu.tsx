import React from "react";
import EventListener from "react-event-listener";
import styled, { css } from "styled-components";

interface IMenuItemProps extends React.HTMLAttributes<{}> {
  // TODO: HTMLタグリストを示す型を利用する
  as?: React.ComponentType | "li" | "button" | "div" | "a";
  style?: React.CSSProperties;
  disabled?: boolean;
  component?: React.ElementType;
}

/**
 * メニューアイテムコンポーネント
 *
 * @param button ボタンコンポーネントを使用するか  true: ボタンコンポーネントを利用
 * @param to     リンク先のパス
 */
const MenuItem: React.SFC<IMenuItemProps> = ({
  as = "div",
  style = null,
  disabled = false,
  children,
  ...otherProps
}) => {
  return (
    <MenuItemStyled as={as} style={style} disabled={disabled} {...otherProps}>
      {children}
    </MenuItemStyled>
  );
};

interface IMenuProps {
  title: string;
  open?: boolean;
  onClose?: (e: MouseEvent) => void;
}

const Menu: React.SFC<IMenuProps> = ({
  title,
  children,
  open = false,
  onClose,
  ...otherProps
}) => {
  if (!open) {
    return null;
  }
  return (
    <MenuMain>
      <EventListener target={document} onClick={e => onClose(e)} />
      <MenuItemStyled>{title}</MenuItemStyled>
      <MenuList>{children}</MenuList>
    </MenuMain>
  );
};

export { Menu, MenuItem, IMenuItemProps, MenuList, IMenuProps };

const MenuMain = styled.div`
  position: absolute;
  min-width: 16px;
  min-height: 16px;
  overflow-y: auto;
  overflow-x: hidden;
  border-radius: 2px;
  background-color: #fff;
  opacity: 1;
  transform: scale(1, 1) translate(-50px, 57px);
  /* transform-origin: 0px 32px; */
  transition: opacity 267ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    transform 178ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  box-shadow: 0px 5px 5px -3px rgba(0, 0, 0, 0.2),
    0px 8px 10px 1px rgba(0, 0, 0, 0.14), 0px 3px 14px 2px rgba(0, 0, 0, 0.12);
`;

const MenuList = styled.ul`
  padding-top: 8px;
  padding-bottom: 8px;
`;

const disabledStyle = css`
  cursor: default;
  &:hover {
    opacity: 1;
  }
`;

const MenuItemStyled = styled.div<IMenuItemProps>`
  display: block;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 12px;
  padding-bottom: 12px;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5em;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: pointer;

  &:hover {
    opacity: 0.7;
  }

  ${({ disabled }) => (disabled ? disabledStyle : "")};
`;
