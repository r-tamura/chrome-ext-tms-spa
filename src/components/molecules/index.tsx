export * from "./Menu";
export * from "./NavItem";
export * from "./CancelOkButtons";
