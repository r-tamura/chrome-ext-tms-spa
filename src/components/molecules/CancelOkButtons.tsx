import React from "react";
import { ButtonGroup, Button } from "../atoms";

interface IProps {
  handleCancelClick(e: React.MouseEvent<HTMLButtonElement>): void;
  isCancelDisabled?: boolean;
  cancelText?: string;
  isSubmitDisabled?: boolean;
  submitText?: string;
}

export function CancelOkButtons({
  handleCancelClick,
  isCancelDisabled = false,
  isSubmitDisabled = false,
  cancelText = "キャンセル",
  submitText = "サブミット"
}: IProps) {
  return (
    <ButtonGroup>
      <Button
        variant="contained"
        type="button"
        onClick={handleCancelClick}
        disabled={isCancelDisabled}
      >
        {cancelText}
      </Button>
      <Button variant="contained" color="primary" disabled={isSubmitDisabled}>
        {submitText}
      </Button>
    </ButtonGroup>
  );
}
