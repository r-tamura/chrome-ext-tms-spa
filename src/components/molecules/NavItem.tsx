import React, { useContext } from "react";
import { Link } from "react-router-dom";
import { MenuItem } from "./Menu";
import styled, { ThemeContext } from "styled-components";

interface INavProps {
  to: string; // Routeのナビゲーション先
  selected?: boolean; // 選択中であるか
  disabled?: boolean; // 無効リンクであるか
}

/**
 * ナビゲーションアイテムコンポーネント
 * サイドナビゲーションの項目
 *
 * @param {string} to Routeのナビゲーション先
 * @param {boolean} selected 選択中であるか(true: 選択中/false: 未選択)
 * @param {boolean} disabled 無効なリンクであるか(true: 無効/false: 有効)
 * @param {React.ReactNode} children 子コンポーネント
 */
export const NavItem: React.SFC<INavProps> = ({
  to,
  disabled = false,
  selected = false,
  children
}) => {
  const theme = useContext(ThemeContext);
  const StyledLink = disabled ? DisabeldLink : BaseLink;

  return (
    <MenuItem
      disabled={disabled}
      style={
        selected
          ? {
              backgroundColor: theme.primaryDark
            }
          : null
      }
    >
      <StyledLink to={to}>{children}</StyledLink>
    </MenuItem>
  );
};

const BaseLink = styled(Link)`
  color: ${({ theme }) => theme.white};
  display: block;
`;

const DisabeldLink = styled.div`
  color: ${({ theme }) => theme.white};
  opacity: 0.7;
`;
