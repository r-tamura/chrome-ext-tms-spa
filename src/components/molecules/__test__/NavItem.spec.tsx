import React from "react";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router";
import toJson from "enzyme-to-json";
import { matchers } from "jest-emotion";
import { NavItem } from "~/components/molecules";
import { AppThemeProvider } from "~/components/Provider";
import { telema } from "~/styles/theme";

expect.extend(matchers);

const withRouter = (children: React.ReactNode) => (
  // keyを与えないとスナップショットテストに失敗する
  // refers: https://github.com/ReactTraining/react-router/issues/5579
  <AppThemeProvider>
    <MemoryRouter initialEntries={[{ pathname: "/", key: "testKey" }]}>
      {children}
    </MemoryRouter>
  </AppThemeProvider>
);

describe("<NavItem />", () => {
  it("Snapshot", () => {
    const $ = mount(withRouter(<NavItem to="/">Item</NavItem>));

    expect(toJson($)).toMatchSnapshot();
  });

  describe("未選択状態", () => {
    const $ = mount(withRouter(<NavItem to="/">Item</NavItem>));
    const $link = $.find("a");

    it("未選択状態", () => {
      const $Item = $.find("div");
      expect($Item.length).toBe(1);
      expect($Item.prop("disabled")).toBe(false);
    });

    it("指定されたテキストが表示される", () => {
      expect($.text()).toBe("Item");
    });
    it("リンクタグを含む", () => {
      expect($link.length).toBe(1);
    });

    it("リンク先が指定されたパスとなっている", () => {
      expect($link.prop("href")).toBe("/");
    });

    // TODO: emotionのstyled-componentsで指定されたスタイルがテストできない
    // it("リンクはblockスタイル", () => {
    //   // console.log(toJson($link));
    //   expect(toJson($)).toHaveStyleRule("display", "block");
    // });
  });

  describe("選択状態", () => {
    const $ = mount(
      withRouter(
        <NavItem to="/" selected={true}>
          Item
        </NavItem>
      )
    );
    it("背景色がプライマリカラーの濃い色", () => {
      const $item = $.find("div");
      const bg = ($item.getDOMNode() as HTMLElement).style.backgroundColor;
      expect(bg).toBe(telema.primaryDark);
    });
  });

  describe("無効状態", () => {
    const $ = mount(
      withRouter(
        <NavItem to="/" disabled={true}>
          Item
        </NavItem>
      )
    );

    it("リンクが無効(aタグが存在しない)", () => {
      const $link = $.find("a");
      expect($link.length).toBe(0);
    });
  });
  // it("should be disabled", () => {

  //   expect(w.is("div")).toBe(true);
  //   expect(w.hasClass("disabled")).toBeTruthy();
  // });

  // it("static rendering", () => {
  //   const w = mount(withRouter(<NavItem to="/">Item</NavItem>));
  //   expect(w.is("a")).toBe(true);
  //   expect(w.hasClass("selected")).toBeFalsy();
  //   expect(w.hasClass("disabled")).toBeFalsy();
  // });
});
