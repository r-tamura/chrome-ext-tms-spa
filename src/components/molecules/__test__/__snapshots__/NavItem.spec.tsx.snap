// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavItem /> Snapshot 1`] = `
<Component>
  <ThemeProvider
    theme={
      Object {
        "black": "rgb(5, 5, 5)",
        "buttonDefault": "rgba(0, 0, 0, 0.08)",
        "danger": "#D32F2F",
        "fontFamily": "\\"M PLUS Rounded 1c\\", sans-serif",
        "formBorderColor": "rgba(0, 0, 0, 0.16)",
        "formBorderColorDisabled": "rgba(0, 0, 0, 0.08)",
        "formBorderColorError": "#D32F2F",
        "formBorderColorFocused": "hsl(206, 100%, 31%)",
        "hoveredBg": "rgba(0, 0, 0, 0.03)",
        "inputPlaceHolder": "rgba(0, 0, 0, 0.09)",
        "primary": "hsl(206, 100%, 31%)",
        "primaryDark": "hsl(208, 80%, 27%)",
        "primaryHeavyDark": "hsl(208, 80%, 27%)",
        "primaryHeavyLight": "hsl(199, 91%, 64%);",
        "primaryLight": "hsl(201, 98%, 41%);",
        "primaryReverse": "hsla(0, 0%, 99%, 1);",
        "primaryReverseDark": "hsla(0, 0%, 99%, 0.5);",
        "secondary": "#689F38",
        "textGray": "hsla(0, 5%, 5%, .25)",
        "textMain": "hsl(0, 5%, 5%)",
        "white": "rgb(253, 253, 253)",
      }
    }
  >
    <MemoryRouter
      initialEntries={
        Array [
          Object {
            "key": "testKey",
            "pathname": "/",
          },
        ]
      }
    >
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "testKey",
                "pathname": "/",
                "search": "",
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <WithTheme(NavItemBase)
          to="/"
        >
          <NavItemBase
            theme={
              Object {
                "black": "rgb(5, 5, 5)",
                "buttonDefault": "rgba(0, 0, 0, 0.08)",
                "danger": "#D32F2F",
                "fontFamily": "\\"M PLUS Rounded 1c\\", sans-serif",
                "formBorderColor": "rgba(0, 0, 0, 0.16)",
                "formBorderColorDisabled": "rgba(0, 0, 0, 0.08)",
                "formBorderColorError": "#D32F2F",
                "formBorderColorFocused": "hsl(206, 100%, 31%)",
                "hoveredBg": "rgba(0, 0, 0, 0.03)",
                "inputPlaceHolder": "rgba(0, 0, 0, 0.09)",
                "primary": "hsl(206, 100%, 31%)",
                "primaryDark": "hsl(208, 80%, 27%)",
                "primaryHeavyDark": "hsl(208, 80%, 27%)",
                "primaryHeavyLight": "hsl(199, 91%, 64%);",
                "primaryLight": "hsl(201, 98%, 41%);",
                "primaryReverse": "hsla(0, 0%, 99%, 1);",
                "primaryReverseDark": "hsla(0, 0%, 99%, 0.5);",
                "secondary": "#689F38",
                "textGray": "hsla(0, 5%, 5%, .25)",
                "textMain": "hsl(0, 5%, 5%)",
                "white": "rgb(253, 253, 253)",
              }
            }
            to="/"
          >
            <MenuItem
              disabled={false}
              style={null}
            >
              <styled.div
                as="div"
                disabled={false}
                style={null}
              >
                <StyledComponent
                  as="div"
                  disabled={false}
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-htpNat",
                        "isStatic": false,
                        "lastClassName": "XnzGZ",
                        "rules": Array [
                          "
  display: block;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 12px;
  padding-bottom: 12px;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5em;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: pointer;

  &:hover {
    opacity: 0.7;
  }

  ",
                          [Function],
                          ";
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-htpNat",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  style={null}
                >
                  <div
                    className="sc-htpNat XnzGZ"
                    disabled={false}
                    style={null}
                  >
                    <Styled(Link)
                      to="/"
                    >
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bxivhb",
                              "isStatic": false,
                              "lastClassName": "gjOGIy",
                              "rules": Array [
                                "
  color: ",
                                [Function],
                                ";
  display: block;
",
                              ],
                            },
                            "displayName": "Styled(Link)",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bxivhb",
                            "target": [Function],
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        to="/"
                      >
                        <Link
                          className="sc-bxivhb gjOGIy"
                          to="/"
                        >
                          <a
                            className="sc-bxivhb gjOGIy"
                            href="/"
                            onClick={[Function]}
                          >
                            Item
                          </a>
                        </Link>
                      </StyledComponent>
                    </Styled(Link)>
                  </div>
                </StyledComponent>
              </styled.div>
            </MenuItem>
          </NavItemBase>
        </WithTheme(NavItemBase)>
      </Router>
    </MemoryRouter>
  </ThemeProvider>
</Component>
`;
